Client: Remove core dependancy from the client

haozturk opened the pull request:
Fixes #7746 

In particular, removing `(rucio.db.sqla.constants.TransferLimitDirection)` from the `requestclient.py`. This removes the ability to set the transfer limit direction via the client. Other abilities are preserved, e.g. [1]. As an alternative, we could hard code the type annotation as well (`direction: Union['D', 'S'] = 'D'`), but this is simply not maintainable long-term.


@dchristidis @bari12 FYI

[1]
```
>>> from rucio.client import Client
>>> client = Client()
>>> print(list(client.list_transfer_limits()))
[]
>>> print(client.set_transfer_limit(rse_expression="MOCK", max_transfers=21))
True
>>> print(list(client.list_transfer_limits()))
[{'activity': 'all_activities', 'rse_expression': 'MOCK', 'direction': 'DESTINATION', 'volume': 0, 'strategy': 'fifo', 'waitings': None, 'updated_at': datetime.datetime(2025, 5, 27, 8, 54), 'id': 'a655c177d7b245089075247360d96237', 'max_transfers': 21, 'deadline': 1, 'transfers': None, 'created_at': datetime.datetime(2025, 5, 27, 8, 54)}]
>>> print(client.delete_transfer_limit(rse_expression="MOCK"))
True
>>> print(list(client.list_transfer_limits()))
[]
```

haozturk committed:
Client: Remove core dependancy(rucio.db.sqla.constants.TransferLimitDirection) from the client

rdimaio commented:
I think we should move `TransferLimitDirection` to `rucio.common.constants`, by virtue of the fact that it is now needed outside of core (and thus, it qualifies as 'common'); this way, we can also preserve the ability to set the transfer limit direction via the client; @dchristidis @bari12 thoughts?

bari12 commented:
I think that's a good idea.

haozturk commented:
Okay, I'll update the merge the request accordingly then. I didn't know that this was an option. It's indeed better.

haozturk commented:
I hit into a circular import issue while doing this. Converting `TransferLimitDirection` import in `rucio.db.sqla.models` [1] from 
```
from rucio.db.sqla.constants import TransferLimitDirection
```
to
```
from rucio.common.constants import TransferLimitDirection
```
yields this error [2] because `rucio.common.constants` depends on `rucio.db.sqla.models` [3] (Config table to be specific).

Any suggestion on how to move on? I'm inclined to say that `rucio.common.constants` shouldn't depend on any internal module. Said that, I might be wrong and I'm not sure where the config dependent mapping can be moved.

[1]https://github.com/rucio/rucio/blob/a790545443208016bf808eacf97e682349e769b7/lib/rucio/db/sqla/models.py#L54
[2]
```
bash-5.1# tools/run_tests.sh -i
Clearing memcache
tools/run_tests.sh: connect: Connection refused
tools/run_tests.sh: line 66: /dev/tcp/127.0.0.1/11211: Connection refused
Cleaning *.pyc files
Cleaning old authentication tokens
Cleaning local RSE directories
Using the standard config
Resetting database tables
Traceback (most recent call last):
  File "/opt/rucio/tools/reset_database.py", line 25, in <module>
    from rucio.db.sqla.util import build_database, create_base_vo, create_root_account, destroy_database, drop_everything  # noqa: E402
  File "/usr/local/lib/python3.9/site-packages/rucio/db/sqla/util.py", line 39, in <module>
    from rucio.db.sqla import models
  File "/usr/local/lib/python3.9/site-packages/rucio/db/sqla/models.py", line 33, in <module>
    from rucio.db.sqla.constants import (
ImportError: cannot import name 'TransferLimitDirection' from 'rucio.db.sqla.constants' (/usr/local/lib/python3.9/site-packages/rucio/db/sqla/constants.py)
Failed to reset the database!
bash-5.1# tools/run_tests.sh -i
Clearing memcache
Cleaning *.pyc files
Cleaning old authentication tokens
Cleaning local RSE directories
Using the standard config
Resetting database tables
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/rucio/common/config.py", line 193, in config_get
    return convert_type_fnc(get_config().get(section, option))
  File "/usr/lib64/python3.9/configparser.py", line 781, in get
    d = self._unify_values(section, vars)
  File "/usr/lib64/python3.9/configparser.py", line 1152, in _unify_values
    raise NoSectionError(section) from None
configparser.NoSectionError: No section: 'transfers'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/rucio/tools/reset_database.py", line 25, in <module>
    from rucio.db.sqla.util import build_database, create_base_vo, create_root_account, destroy_database, drop_everything  # noqa: E402
  File "/usr/local/lib/python3.9/site-packages/rucio/db/sqla/util.py", line 39, in <module>
    from rucio.db.sqla import models
  File "/usr/local/lib/python3.9/site-packages/rucio/db/sqla/models.py", line 32, in <module>
    from rucio.common.constants import TransferLimitDirection
  File "/usr/local/lib/python3.9/site-packages/rucio/common/constants.py", line 45, in <module>
    if config_get_bool('transfers', 'srm_https_compatibility', raise_exception=False, default=False):
  File "/usr/local/lib/python3.9/site-packages/rucio/common/config.py", line 493, in config_get_bool
    return config_get(
  File "/usr/local/lib/python3.9/site-packages/rucio/common/config.py", line 201, in config_get
    return __config_get_table(section=section, option=option, raise_exception=raise_exception,
  File "/usr/local/lib/python3.9/site-packages/rucio/common/config.py", line 637, in __config_get_table
    return core_config_get(section, option, default=default, session=session, use_cache=use_cache,
  File "/usr/local/lib/python3.9/site-packages/rucio/common/utils.py", line 1654, in _wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rucio/db/sqla/session.py", line 402, in new_funct
    return function(*args, session=session, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/rucio/core/config.py", line 210, in get
    models.Config.value
AttributeError: partially initialized module 'rucio.db.sqla.models' has no attribute 'Config' (most likely due to a circular import)
Failed to reset the database!
```
[3] https://github.com/rucio/rucio/blob/a790545443208016bf808eacf97e682349e769b7/lib/rucio/common/constants.py#L19

rdimaio commented:
> I'm inclined to say that rucio.common.constants shouldn't depend on any internal module.

I agree with you. It looks like there's a dependency on `rucio.common.config` which is causing the issue. This is only used to populate the `SCHEME_MAP`. I will restructure this a bit so that this circular dependency is avoided under this issue https://github.com/rucio/rucio/issues/7758

haozturk commented:
Thanks Riccardo, I will follow your PR and make the necessary changes when it's merged.

rdimaio commented:
> Thanks Riccardo, I will follow your PR and make the necessary changes when it's merged.

Thanks for the patience - https://github.com/rucio/rucio/pull/7759 has been merged, you can rebase and it should work

bari12 commented:
@haozturk Riccardo's PR is merged. Please update this one, I would like to include this in the upcoming release. (Should be today, but I will move it to tomorrow)

haozturk commented:
Sure, I'll update it today. We received multiple complaints in CMS that the client 37.4.0 is broken, so indeed better to release  a fix sooner than later.

haozturk commented:
I've made a new PR where I moved `TransferLimitDirection` from `rucio.db.sqla.constants` to `rucio.common.constants` and updated all its imports: https://github.com/rucio/rucio/pull/7766

Feel free to close this one. I created a separate PR just in case the new PR isn't ready to be merged and you prefer this one to go in to fix the client 37 release although it's not the ideal implementation (after rebasing)

rdimaio commented:
> I've made a new PR where I moved `TransferLimitDirection` from `rucio.db.sqla.constants` to `rucio.common.constants` and updated all its imports: #7766
> 
> Feel free to close this one. I created a separate PR just in case the new PR isn't ready to be merged and you prefer this one to go in to fix the client 37 release although it's not the ideal implementation (after rebasing)

Closed in favour of https://github.com/rucio/rucio/pull/7766
